### PR TITLE
layers: GH1720 Migrate swapchain queue family checks 

### LINKED ISSUE #1720: Layers are too restrictive with gotQueueFamilyPropertyCount for vkGetPhysicalDeviceSurfaceSupportKHR
In validation layers there is this [condition](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/572ea79/layers/swapchain.cpp#L266) (simplified here):

`if( qfp_count && !qfp_is_null ) gotQueueFamilyPropertyCount = true;`

I think this is over-restrictive. E.g. this (IMO valid and reasonable) made up code would generate an validation error:

    bool fun( VkPhysicalDevice physicalDevice, VkSurfaceKHR surface ){
        uint32_t qfCount;
        vkGetPhysicalDeviceQueueFamilyProperties( physicalDevice, &qfCount, nullptr );

        if( qfCount > 0 ){
            VkBool32 isSupported;
            VkResult errorCode = vkGetPhysicalDeviceSurfaceSupportKHR(
                physicalDevice, queueFamily, surface, &isSupported
            );

            return isSupported;
        }
        
       return false;
    }

> ERROR: PhysicalDevice(0x17523E6A540): 24, Swapchain, vkGetPhysicalDeviceSurfaceSupportKHR() called before calling the vkGetPhysicalDeviceQueueFamilyProperties function.

<hr/>

I can PR this myself, given time.

<hr/>

Also the `phy_data->numOfQueueFamilies` seems to be rewritten every time on `vkGetPhysicalDeviceQueueFamilyProperties`.

And `phy_data->gotQueueFamilyPropertyCount` seems to be reset every `vkEnumeratePhysicalDevices`.

That gives an Error for many valid usages of Vulkan, and I suspect for some "reasonable" usages too,

### PR DESCRIPTION
**Pull request by krOoze**
continuing PR #1738 
based on PR #1742 updates (identical commit 1).

 - Fix #1720
 - Help with #1683 &mdash; migrate queue family checks from swapchain layer into core or param validatiom.
 - Check queue family uniqueness in array
 - Add some checks applicable to `vkCreateSwapchainKHR` (copied from `vkCreateImage`)

### CODE REVIEW

**Comment by chrisforbes** [comment]
@krOoze thanks for taking on all of this cleanup! #1683 is getting pretty close to done :)

**Comment by mark-lunarg** [comment]
@krOoze, I rebased on master and was going to push, but noticed a couple of new LayerValidationTest failures (See the vk_layer_validaiton_tests window subproject, or tests/layer_validation_tests.cpp) -- these tests may need to be updated to look for warnings or for message text changing, or perhaps the tests are now doing something wrong -- in any case, the unexpected errors should be fixed in the tests.  If this is too much trouble let me know and I'll take a stab at it.  Failures are on Linux and Windows as follows:
```
[ RUN      ] VkLayerTest.InvalidQueueFamilyIndex
Unexpected: vkCreateBuffer: pCreateInfo->pQueueFamilyIndices[0] (=777) is not one of the queue families given via VkDeviceQueueCreateInfo structures when the device was created.  [00001]
Other error messages logged for this test were:
     vkCreateBuffer: pCreateInfo->pQueueFamilyIndices[0] (=777) is not one of the queue families given via VkDeviceQueueCreateInfo structures when the device was created.  [00001]
C:\dev\xgl\Vulkan-LoaderAndValidationLayers\tests\layer_validation_tests.cpp(291): error: Failed
Did not receive expected error 'vkCreateBuffer: pCreateInfo->pQueueFamilyIndices[0] (777) must be one of the indices specified when the device was created, via the VkDeviceQueueCreateInfo structure.'
[  FAILED  ] VkLayerTest.InvalidQueueFamilyIndex (733 ms)
```
 <and> 

```
[ RUN      ] VkLayerTest.MismatchCountQueueCreateRequestedFeature
Unexpected: vkCreateDevice: pCreateInfo->pQueueCreateInfos[0].queueFamilyIndex is VK_QUEUE_FAMILY_IGNORED, but it is required to provide a valid queue family index value. [00004]
Unexpected: vkCreateDevice: pCreateInfo->pQueueCreateInfos[0].queueFamilyIndex (=4294967295) is not less than any previously obtained pQueueFamilyPropertyCount from vkGetPhysicalDeviceQueueFamilyProperties (i.e. is not less than 2). For more information refer to Vulkan Spec Section '4.3.2. Queue Creation' which states 'queueFamilyIndex must be less than pQueueFamilyPropertyCount returned by vkGetPhysicalDeviceQueueFamilyProperties' https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#VkDeviceQueueCreateInfo) [00054]
Other error messages logged for this test were:
     vkCreateDevice: pCreateInfo->pQueueCreateInfos[0].queueFamilyIndex is VK_QUEUE_FAMILY_IGNORED, but it is required to provide a valid queue family index value. [00004]
     vkCreateDevice: pCreateInfo->pQueueCreateInfos[0].queueFamilyIndex (=4294967295) is not less than any previously obtained pQueueFamilyPropertyCount from vkGetPhysicalDeviceQueueFamilyProperties (i.e. is not less than 2). For more infor
mation refer to Vulkan Spec Section '4.3.2. Queue Creation' which states 'queueF
amilyIndex must be less than pQueueFamilyPropertyCount returned by vkGetPhysical
DeviceQueueFamilyProperties' (https://www.khronos.org/registry/vulkan/specs/1.0-
extensions/html/vkspec.html#VkDeviceQueueCreateInfo) [00054]
C:\dev\xgl\Vulkan-LoaderAndValidationLayers\tests\layer_validation_tests.cpp(291): error: Failed
Did not receive expected error 'Invalid queue create request in vkCreateDevice()
. Invalid queueFamilyIndex '
[  FAILED  ] VkLayerTest.MismatchCountQueueCreateRequestedFeature (853 ms)
```

So close!

**Comment by krOoze** [comment]
@mark-lunarg Sure I will look into this in a bit...

**Comment by krOoze** [comment]
@mark-lunarg Rebased. Seems fine too.

**Comment by mark-lunarg** [comment]
@krOoze, we won't be able to push this until the test failures are fixed.  I don't think the CI checks run this test, but as a bare minimum it must pass after each individual commit in any PR. If you like, I can amend you commits with the necessary fixes for this PR and we can go from there.

**Comment by krOoze** [comment]
@mark-lunarg I am not getting any test failures (except the preexisting one):
https://gist.github.com/krOoze/90a4a65f0b2588e144a2725be6e7d241

**Comment by mark-lunarg** [comment]
@krOoze, Ah, super.  Much appreciation for running the test and checking things out.  Can you share what platform(s) you are running on?  I'd like to get your failure fixed as well.

Thanks again for all this awesome work!

**Comment by krOoze** [comment]
@mark-lunarg I suspect though I don't get all tests. Is there really only 38 of them?

Anyway, platform:
Windows 10 Pro N 1703 x64
Intel i5-3340
AMD 280X, software 17.5.1
x64 build version of the tests

**Comment by mark-lunarg** [comment]
Oh, I see, those are the Loader Validation Tests.  So, the run_all_tests script will do some test layer magic and then run the layer validation tests a couple of times, but after those all run it should then execute the layer validation tests.

You can also easily run standalone test from within Visual Studio by just setting the 
`vk_layer_validation_tests `project as the 'StartUp Project' and hitting Start Debug or Start Without Debugging.  Lemme know...

**Comment by krOoze** [comment]
@mark-lunarg Aha, now I am getting some juicy failures. So the `run_all_tests` does not work then...

**Comment by krOoze** [comment]
@mark-lunarg Tests OK now. Though I get 2 preexisting failures: https://gist.github.com/krOoze/647bb06a27fae0101ea4e18ac4fdd325

File: tests/layer_validation_tests.cpp [hunk 1]
```diff
@@ -10089,7 +10089,7 @@ TEST_F(VkLayerTest, InvalidQueueFamilyIndex) {
         vkFreeMemory(m_device->device(), mem, NULL);
     }
 
-    vkDestroyBuffer(m_device->device(), ib, NULL);
+    if(!err_ib1) vkDestroyBuffer(m_device->device(), ib, NULL);
```
Comment (krOoze) [inline tests/layer_validation_tests.cpp]: Though do not understand how this was passing before.... the buffer potentially failed creation, right?

**Comment by mark-lunarg** [comment]
@krOoze, pushed to Master!  I added a commit that removed the destroybuffer call -- if the test fails it won't matter if there's another unexpected error.  Thanks again -- hopefully the next ones will be much easier.

**Pull request closed without merging**